pkg/api/steve/supportconfigs/generator.go

```diff
@@ -0,0 +1,168 @@
+	archive, err := h.generateSupportConfig()
+	logrus.Infof("[%s] Done Generating supportconfig", logPrefix)
+	if err != nil {
+		logrus.Errorf("[%s] Error when generating support config %v", logPrefix, err)
+		_, _ = writer.Write([]byte("Internal Server Error"))
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	writer.Header().Set("Content-Type", tarContentType)
+	writer.Header().Set("Content-Disposition", "attachment; filename=\"supportConfig.tar\"")
+	n, err := io.Copy(writer, archive)
+	logrus.Debugf("[%s] wrote %v bytes in archive response", logPrefix, n)
+}
+
+// authorize checks to see if the user can get the non-resource url of Endpoint. If not, returns an error, otherwise nil
+func (h *GeneratorHandler) authorize(r *http.Request) (bool, error) {
+	userInfo, ok := request.UserFrom(r.Context())
+	if !ok {
+		return false, fmt.Errorf("unable to extract user info from context")
+	}
+	response, err := h.SubjectAccessReviews.Create(r.Context(), &authzv1.SubjectAccessReview{
+		Spec: authzv1.SubjectAccessReviewSpec{
+			// This auth method prevents collision with future k8s resources and allows this to be added to custom rules
+			NonResourceAttributes: &authzv1.NonResourceAttributes{
+				Verb: "get",
+				Path: Endpoint,
+			},
+			User:   userInfo.GetName(),
+			Groups: userInfo.GetGroups(),
+			Extra:  convertExtra(userInfo.GetExtra()),
+			UID:    userInfo.GetUID(),
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		return false, fmt.Errorf("failed to create sar %s", err)
+	}
+	if !response.Status.Allowed {
+		return false, nil
+	}
+	return true, nil
+}
+
+// generateSupportConfig produces an io.Reader with the supportConfig ready to be returned using a http.ResponseWriter
+// Returns an err if it can't get the support config (note: does not return an err if it can't find the csp config)
+func (h *GeneratorHandler) generateSupportConfig() (io.Reader, error) {
+	mergedConfig, err := h.ConfigHandler.getSupportConfigAsMap()
+	if err != nil {
+		return nil, err
+	}
+	cspConfig, err := h.getCspConfig()
+	// csp adapter won't always be installed
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	// merge is only one layer deep - values in the csp adapter will completely override what rancher produces - nested
+	// fields will not be merged
+	for key, value := range cspConfig {
+		// prefer csp's values - CSP config can see rancher's values and the Cloud Providers
+		mergedConfig[key] = value
+	}
+	configData, err := json.MarshalIndent(mergedConfig, "", "  ")
+	if err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+	tw := tar.NewWriter(&buf)
+
+	files := map[string]string{
+		"rancher/config.json": string(configData),
+	}
+
+	for name, body := range files {
+		hdr := &tar.Header{
+			Name: name,
+			Mode: 0600,
+			Size: int64(len(body)),
+		}
+		if err := tw.WriteHeader(hdr); err != nil {
+			return nil, err
+		}
+		if _, err := tw.Write(configData); err != nil {
```

Comment:
Good catch - I think that this actually would have included a bug if we started to do multiple files (every file would have just had the config data as the value instead of the body as intended). I've fixed accordingly to go along with your suggestion.